File: rust/0139_word_break/bfs.rs
/*
139. Word Break
Medium

Given a non-empty string s and a dictionary wordDict containing a list of non-empty words, determine if s can be segmented into a space-separated sequence of one or more dictionary words.

Note:

    The same word in the dictionary may be reused multiple times in the segmentation.
    You may assume the dictionary does not contain duplicate words.

Example 1:

Input: s = "leetcode", wordDict = ["leet", "code"]
Output: true
Explanation: Return true because "leetcode" can be segmented as "leet code".

Example 2:

Input: s = "applepenapple", wordDict = ["apple", "pen"]
Output: true
Explanation: Return true because "applepenapple" can be segmented as "apple pen apple".
             Note that you are allowed to reuse a dictionary word.

Example 3:

Input: s = "catsandog", wordDict = ["cats", "dog", "sand", "and", "cat"]
Output: false


https://leetcode.com/problems/word-break/submissions/
*/
use std::collections::{HashSet, VecDeque};
impl Solution {
    pub fn word_break(s: String, word_dict: Vec<String>) -> bool {
        let mut queue: VecDeque<usize> = VecDeque::new();
        let mut visited: HashSet<usize> = HashSet::new();
        let words = word_dict.into_iter().collect::<HashSet<String>>();
        queue.push_back(0);
        visited.insert(0);
        while queue.len() > 0 {
            let a = queue.pop_front().unwrap();

            for b in a + 1..s.len() + 1 {
                if visited.contains(&b) {
                    continue;
                }
                let maybe = String::from(&s[a..b]);
                if words.contains(&maybe) {
                    if b == s.len() {
                        return true;
                    }
                    queue.push_back(b);
                    visited.insert(b);
                }
            }
        }
        false
    }
}

File: python/0127_word_ladder/bfs.py
"""
127. Word Ladder
Medium

Given two words (beginWord and endWord), and a dictionary's word list, find the length of shortest transformation sequence from beginWord to endWord, such that:

    Only one letter can be changed at a time.
    Each transformed word must exist in the word list.

Note:

    Return 0 if there is no such transformation sequence.
    All words have the same length.
    All words contain only lowercase alphabetic characters.
    You may assume no duplicates in the word list.
    You may assume beginWord and endWord are non-empty and are not the same.

Example 1:

Input:
beginWord = "hit",
endWord = "cog",
wordList = ["hot","dot","dog","lot","log","cog"]

Output: 5

Explanation: As one shortest transformation is "hit" -> "hot" -> "dot" -> "dog" -> "cog",
return its length 5.

Example 2:

Input:
beginWord = "hit"
endWord = "cog"
wordList = ["hot","dot","dog","lot","log"]

Output: 0

Explanation: The endWord "cog" is not in wordList, therefore no possible transformation.
	
https://leetcode.com/problems/word-ladder/
"""
class Solution:
    def ladderLength(self, beginWord: str, endWord: str, wordList: List[str]) -> int:
        from collections import deque, defaultdict
        trans = defaultdict(list)
        L = len(beginWord)
        if endWord not in wordList:
            return 0
        words = [beginWord] + wordList
        for w in words:
            for i in range(L):
                inter = w[:i]+"*"+w[i+1:]
                trans[inter].append(w)
        queue = deque([(beginWord,1)])
        while queue:
            word, lvl = queue.popleft()
            for i in range(L):
                inter = word[:i]+"*"+word[i+1:]
                
                for w in trans[inter]:
                    if w == endWord:
                        return lvl + 1
                    queue.append((w, lvl+1))
                    
                trans[inter] = []
        return 0
        
